docs: add slurmctld high availability explanation (#72)
charmed-hpc/docs: howto/setup/deploy-slurm.md

:::::

(deploy-slurmctld-high-availability)=
### `slurmctld` high availability

The `slurmcltd` charm supports [high availability (HA)](explanation-high-availability) through the native functionality provided by Slurm: an active-passive setup where additional units are backups to a single primary.

> This functionality requires a low-latency [shared file system to be deployed](howto-setup-deploy-shared-filesystem) and a `filesystem-client` charm to be integrated with `slurmctld` on the `mount` endpoint to allow sharing of data across all `slurmctld` units. It is recommended that this file system be separate from the file system used for cluster compute nodes to avoid I/O-intensive user jobs from impacting `slurmctld` responsiveness.

Reviewer: There's a statement in the explanation noting that the `filesystem-client` must be deployed without a user-configured mount point - would this be an appropriate place within the how-to to reiterate this point?

    endpoint = module.filesystem-client.requires.filesystem
  }
}

module "slurmctld" {
  source      = "git::https://github.com/charmed-hpc/slurm-charms//charms/slurmctld/terraform"
  model_name  = juju_model.slurm.name
  constraints = "arch=amd64 virt-type=virtual-machine"
  units       = 2
}

resource "juju_integration" "filesystem-to-slurmctld" {
  model = juju_model.slurm.name

  application {
    name     = module.slurmctld.app_name
    endpoint = module.slurmctld.provides.mount
  }

  application {
    name     = module.filesystem-client.app_name
    endpoint = module.filesystem-client.requires.mount
  }
}
:::

::::

:::::

(migrate-slurmctld-high-availability)=
#### Migrating a single `slurmctld` to high availability

:::{warning}
**This migration requires cluster downtime**.

The `slurmctld` service is stopped during the copy of Slurm data from the unit local `/var/lib/slurm/checkpoint` to shared storage. Downtime varies depending on the scale of data to be transferred and transfer rate to the shared storage.

This is a one-time cluster downtime. Once the data migration is complete, no further downtime is necessary when adding or removing `slurmctld` units.
:::

> To migrate a previously deployed single `slurmctld` unit to an HA setup, integrate a `filesystem-client`, substituting `[filesystem-provider]` with your choice of shared file system provider (file system provider choice is detailed in the [`slurmctld` high availability deployment instructions](deploy-slurmctld-high-availability)). Then scale up the application:

Reviewer: ```suggestion
To migrate a previously deployed single `slurmctld` unit to an HA setup, integrate a `filesystem-client` following the file system deployment process described in the [`slurmctld` high availability section](deploy-slurmctld-high-availability), substituting `[filesystem-provider]` with your choice of shared file system provider. Then scale up the application:
```

::::{tab-item} Terraform
:sync: terraform

:::{code-block} terraform
:caption: `main.tf`
module "filesystem-client" {
  source     = "git::https://github.com/charmed-hpc/filesystem-charms//charms/filesystem-client/terraform"
  model_name  = juju_model.slurm.name
}

resource "juju_integration" "provider-to-filesystem" {
  model = juju_model.slurm.name

  application {
    name     = module.[filesystem-provider].app_name
    endpoint = module.[filesystem-provider].provides.filesystem
  }

  application {
    name     = module.filesystem-client.app_name
    endpoint = module.filesystem-client.requires.filesystem
  }
}

module "slurmctld" {
  source      = "git::https://github.com/charmed-hpc/slurm-charms//charms/slurmctld/terraform"
  model_name  = juju_model.slurm.name
  constraints = "arch=amd64 virt-type=virtual-machine"
  units       = 2
}

resource "juju_integration" "filesystem-to-slurmctld" {
  model = juju_model.slurm.name

  application {
    name     = module.slurmctld.app_name
    endpoint = module.slurmctld.provides.mount
  }

  application {
    name     = module.filesystem-client.app_name
    endpoint = module.filesystem-client.requires.mount
  }
}
:::

::::

:::::

> 

Reviewer: A closing statement and example on what the scaled up system would look like with the appropriate commands would be helpful here

Author: Added this. Since this is a specific example, I included CephFS as the shared file system since that's what we've been using in testing.

I also went back and changed the other `juju status` outputs to include the new `primary - UP` unit message for `slurmcltd`. It's now present even in a non-HA setup.

:::::

(deploy-slurmctld-high-availability)=
### Deploying `slurmctld` in high availability

> The `slurmcltd` charm optionally supports [high availability (HA)](explanation-high-availability) through the native functionality provided by Slurm: an active-passive setup where additional units are backups to a single primary.

Author: I like this one as "controllers" since it refers to native Slurm functionality and not Juju.

> This functionality requires a low-latency [shared file system to be deployed](howto-setup-deploy-shared-filesystem) and a `filesystem-client` charm, without a user-configured mount point, to be integrated with `slurmctld` on the `mount` endpoint to allow sharing of data across all `slurmctld` units. For guidance on choosing a file system, see the [Shared `StateSaveLocation` using `filesystem-client` charm](explanation-slurmctld-high-availability-state-save-location) section.
> 
> It is recommended that the HA file system **not be the same as the file system used for the cluster compute nodes** to avoid I/O-intensive user jobs from impacting `slurmctld` responsiveness. The suggested approach is to deploy a dedicated HA file system then subsequently provision a separate file system for the compute nodes.
> 
> Once a chosen shared file system has been deployed and made available via a proxy or other file system provider charm, run the following, substituting `[filesystem-provider]` with the name of the provider charm, to deploy a `slurmctld` HA setup with two units (a primary and single backup):

Reviewer: ```suggestion
This functionality requires a low-latency shared file system to be deployed and a `filesystem-client` charm to be integrated with `slurmctld`. For guidance on choosing a file system, see the [Shared `StateSaveLocation` using `filesystem-client` charm](explanation-slurmctld-high-availability-state-save-location) section. It is recommended that the HA file system **not be the same as the file system used for the cluster compute nodes** to avoid I/O-intensive user jobs from impacting `slurmctld` responsiveness. The suggested approach is to deploy a dedicated HA file system then subsequently provision a separate file system for the compute nodes.

To deploy the file system, follow the instructions in the [Deploy a shared filesystem](howto-setup-deploy-shared-filesystem) section. Once a chosen shared file system has been deployed and made available via a proxy or provider charm, run the following, substituting `[filesystem-provider]` with the name of the provider charm, to deploy a `slurmctld` HA setup with two units (a primary and single backup):
```

Author: Fixed!

:::::{tab-set}

::::{tab-item} CLI
:sync: cli

:::{code-block} shell
juju deploy filesystem-client --channel latest/edge
juju integrate filesystem-client:filesystem [filesystem-provider]:filesystem

> juju deploy slurmctld --base "ubuntu@24.04" --channel "edge" --num-units 2

Reviewer: What does this `juju deploy slurmctld` do to the already deployed slurmctld controller from above in the main 'Deploy Slurm' section?

Author: Good catch - if there's a `slurmctld` already deployed, doing another `juju deploy slurmctld` would either fail with an "application already exists" error or deploy a brand new application, if either `deploy` used a different name.

I've added a note to the instructions above to read the high availability section instead if HA is needed.

::::{tab-item} Terraform
:sync: terraform

:::{code-block} terraform
:caption: `main.tf`
module "filesystem-client" {
  source     = "git::https://github.com/charmed-hpc/filesystem-charms//charms/filesystem-client/terraform"
  model_name  = juju_model.slurm.name
}

resource "juju_integration" "provider-to-filesystem" {
  model = juju_model.slurm.name

  application {
    name     = module.[filesystem-provider].app_name
    endpoint = module.[filesystem-provider].provides.filesystem
  }

  application {
    name     = module.filesystem-client.app_name
    endpoint = module.filesystem-client.requires.filesystem
  }
}

module "slurmctld" {
  source      = "git::https://github.com/charmed-hpc/slurm-charms//charms/slurmctld/terraform"
  model_name  = juju_model.slurm.name
  constraints = "arch=amd64 virt-type=virtual-machine"
  units       = 2
}

resource "juju_integration" "filesystem-to-slurmctld" {
  model = juju_model.slurm.name

  application {
    name     = module.slurmctld.app_name
    endpoint = module.slurmctld.provides.mount
  }

  application {
    name     = module.filesystem-client.app_name
    endpoint = module.filesystem-client.requires.mount
  }
}
:::

::::

:::::

> 

Reviewer: Might be worth either showing `juju status` here like it's done in the manage section now, or potentially just a quick note to check `juju status` and note the number of `slurmctld` units.

Author: Agreed, I've copied the `juju status` output from the manage section.

:::

Once a chosen shared filesystem has been deployed and made available via a proxy charm, run the following, substituting `filesystem-provider` with the name of the proxy charm, for a `slurmctld` HA setup with two units (a primary and single backup):

:::::{tab-set}

::::{tab-item} CLI
:sync: cli

:::{code-block} shell
juju deploy filesystem-client --channel latest/edge
juju integrate filesystem-client:filesystem filesystem-provider:filesystem

juju deploy slurmctld --base "ubuntu@24.04" --channel "edge" --num-units 2
juju integrate slurmctld:mount filesystem-client:mount
:::

::::

::::{tab-item} Terraform
:sync: terraform

:::{code-block} terraform
:caption: `main.tf`
module "filesystem-client" {
  source     = "git::https://github.com/charmed-hpc/filesystem-charms//charms/filesystem-client/terraform"
  model_name  = juju_model.slurm.name
}

resource "juju_integration" "provider-to-filesystem" {
  model = juju_model.slurm.name

  application {
    name     = module.filesystem-provider.app_name
    endpoint = module.filesystem-provider.provides.filesystem
  }

> 

Author: Unfortunately, \<angle brackets\> in a Terraform block causes issues with the syntax highlighting, giving a `WARNING: Lexing literal_block ... as "terraform" resulted in an error` message that fails the CI tests. We can't use {braces} or \`backticks\` either.

We could use [square brackets] but that'd overlap with the Terraform syntax for a list. I'm considering all-caps: `FILESYSTEM-PROVIDER` but I think it's too inconsistent with the rest of the docs. What do you think?

> 

Author: I've gone with [square brackets].

:::{code-block} shell
juju exec -u sackd/0 -- sinfo
:::

> ## Related how-to guides
> 
> Now that Slurm is deployed, you might want to view:
> 
> - {ref}`howto-setup-deploy-shared-filesystem`

Reviewer 2: We're currently using "Next Steps" to indicate where the user should go next such as in the [COS how-to](https://canonical-charmed-hpc.readthedocs-hosted.com/latest/howto/manage/integrate-with-cos/), so for consistency, I think we should use "Next Steps" here instead

```suggestion
## Next Steps

Now that Slurm is deployed, you might want to view:

- {ref}`howto-setup-deploy-shared-filesystem`
```

The other benefit of using "Next Steps" is that it doesn't limit the scope of where we can tell the user to go next. For example, we could eventually have two paragraphs here, where one is like "here's the next steps for deployment", and the other is "here's some reference documentation to help better understand the architecture of Slurm"

Reviewer 2: Also, not sure how I feel "you might" here for the shared filesystem. It's a must for an actual working HPC cluster :sweat_smile: 

We could say something along the lines here like "now that Slurm is deployed, it is time to deploy the shared filesystem of your Charmed HPC cluster". "You might" here implies to me that this step is optional when it really isn't.
